cmd/influx/secret.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/influxdata/influxdb"
+	"github.com/influxdata/influxdb/cmd/influx/internal"
+	"github.com/influxdata/influxdb/http"
+	"github.com/spf13/cobra"
+	input "github.com/tcnksm/go-input"
+)
+
+type secretSVCsFn func() (influxdb.SecretService, influxdb.OrganizationService, func(*input.UI) string, error)
+
+func cmdSecret(opts ...genericCLIOptFn) *cobra.Command {
+	return newCmdSecretBuilder(newSecretSVCs, opts...).cmd()
+}
+
+type cmdSecretBuilder struct {
+	genericCLIOpts
+
+	svcFn secretSVCsFn
+
+	key string
+	org organization
+}
+
+func newCmdSecretBuilder(svcsFn secretSVCsFn, opts ...genericCLIOptFn) *cmdSecretBuilder {
+	opt := genericCLIOpts{
+		in: os.Stdin,
+		w:  os.Stdout,
+	}
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	return &cmdSecretBuilder{
+		genericCLIOpts: opt,
+		svcFn:          svcsFn,
+	}
+}
+
+func (b *cmdSecretBuilder) cmd() *cobra.Command {
+	cmd := b.newCmd("secret", nil)
+	cmd.Short = "Secret management commands"
+	cmd.Run = seeHelp
+	cmd.AddCommand(
+		b.cmdDelete(),
+		b.cmdFind(),
+		b.cmdUpdate(),
+	)
+	return cmd
+}
+
+func (b *cmdSecretBuilder) cmdUpdate() *cobra.Command {
+	cmd := b.newCmd("update", b.cmdUpdateRunEFn)
+	cmd.Short = "Update secret"
+	cmd.Flags().StringVarP(&b.key, "key", "k", "", "The secret key (required)")
+	cmd.MarkFlagRequired("key")
+	b.org.register(cmd, false)
+
+	return cmd
+}
+
+func (b *cmdSecretBuilder) cmdDelete() *cobra.Command {
+	cmd := b.newCmd("delete", b.cmdDeleteRunEFn)
+	cmd.Short = "Delete secret"
+
+	cmd.Flags().StringVarP(&b.key, "key", "k", "", "The secret key (required)")
+	cmd.MarkFlagRequired("key")
+	b.org.register(cmd, false)
+
+	return cmd
+}
+
+func (b *cmdSecretBuilder) cmdUpdateRunEFn(cmd *cobra.Command, args []string) error {
+	scrSVC, orgSVC, getSecretFn, err := b.svcFn()
+	if err != nil {
+		return err
+	}
+	orgID, err := b.org.getID(orgSVC)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+
+	ui := &input.UI{
+		Writer: b.genericCLIOpts.w,
+		Reader: b.genericCLIOpts.in,
+	}
+	secret := getSecretFn(ui)
+
+	if err := scrSVC.PatchSecrets(ctx, orgID, map[string]string{
+		b.key: secret,
+	}); err != nil {
+		return fmt.Errorf("failed to update secret with key %q: %v", b.key, err)
+	}
+
+	w := internal.NewTabWriter(b.w)
+	w.WriteHeaders("Key", "OrgID", "Updated")
+	w.Write(map[string]interface{}{
+		"Key":     b.key,
+		"OrgID":   orgID,
+		"Updated": true,
+	})
+	w.Flush()
+
+	return nil
+}
+
+func (b *cmdSecretBuilder) cmdDeleteRunEFn(cmd *cobra.Command, args []string) error {
+	scrSVC, orgSVC, _, err := b.svcFn()
+	if err != nil {
+		return err
+	}
+	orgID, err := b.org.getID(orgSVC)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	if err := scrSVC.DeleteSecret(ctx, orgID, b.key); err != nil {
+		return fmt.Errorf("failed to delete secret with key %q: %v", b.key, err)
+	}
+
+	w := internal.NewTabWriter(b.w)
```

Comment:
thought buble just came to me, feels like the `genericCLIOpt` type could be extended with a `tabWriter(headers []string) *tabwriter.Writer` method. That way it always pulls the `b.w` and no one accidentally puts the os.Stdout here. Food for thought 🤔

something like this maybe:

```go
// this exists already
type genericCLIOpts struct {
	in         io.Reader
	w          io.Writer
	runEWrapFn runEWrapFn
}

func (o genericCLIOpts) newTabWriter(headers []string) *internal.TabWriter{
     w := internal.NewTabWriter(b.w)
     w.WriteHeaders(headers)
     return w
}


// secret.go
w := b.newTabWriter("Key", "OrgID", "Deleted")
w.Write(map[string]interface{}{
    "Key":     b.key,
    "OrgID":   orgID,
    "Deleted": true,
})
w.Flush()
```

Reply:
yeah, why not